### PR TITLE
refactor: reduce kmlock deserialization nesting

### PR DESCRIPTION
**Pull request by tykeal**
# Summary
Refactors `_dict_to_kmlocks` to address the coordinator deep-nesting
finding reported in Refs #651 while preserving persisted lock storage
behavior.

## Proposed change
The method now delegates per-field deserialization to focused helpers for
field type resolution, Optional-style Union unwrapping, temporal parsing,
collection dispatch, mapping key coercion, and list conversion. The
commented-out debug statements were kept with the conversion steps they
document.

The mapping branch fall-through behavior is preserved explicitly: once a
field is recognized as mapping-typed, `_kmlock_mapping_value_from_dict`
returns the value for unsupported mapping shapes such as `len(type_args)
!= 2` or non-dict runtime values. Those values therefore do not continue
into the nested dataclass or list conversion branches.

Tests cover temporal parsing and malformed temporal fallback, mapping key
coercion, plain mapping copies, mapping fall-through guards, Union
unwrapping rules, list conversion, non-dataclass fallback, and a
serialize/deserialize lock round trip.

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
- This PR fixes or closes issue:
- This PR is related to issue: Refs #651

### CODE REVIEW

File: custom_components/keymaster/coordinator.py [hunk 1]
```diff
@@ -789,112 +789,181 @@ def _dict_to_kmlocks(self, data: dict, cls: type) -> Any:
...
+        if not isinstance(field_value, dict):
+            return field_value
+
+        # If the value_type is a dataclass, recursively process it
+        if is_dataclass(value_type) and isinstance(value_type, type):
+            # _LOGGER.debug(f"[dict_to_kmlocks] Recursively converting dict items for {field_name}")
+            return {
+                self._kmlock_key_from_dict(k, key_type): self._dict_to_kmlocks(v, value_type)
+                for k, v in field_value.items()
+            }
```
[the remark this answers is not in the document]
Comment (tykeal) [reply custom_components/keymaster/coordinator.py:946]: Thanks for the review. This is intentionally preserved from the pre-refactor implementation: for mapping-typed fields with dataclass values, the old branch called _dict_to_kmlocks(v, value_type) unconditionally for every mapping value. Adding an isinstance(v, dict) guard here would change persisted storage deserialization behavior, while this PR is a depth-only refactor with exact behavior preservation. The non-dict pass-through behavior called out in the PR applies to the mapping field value itself, which is now covered by test_dict_to_kmlocks_mapping_non_dict_value_does_not_fall_through.

**Review by secondof9 — APPROVED** [review]
## 📋 Review Summary
> [!TIP]
> **Review Status:** 🟢 APPROVED
> **Change Type:** 🛠️ Refactor
> **Review Effort:** 🟢 Low
> **Core Impact:** Refactors `KeymasterCoordinator._dict_to_kmlocks` to reduce deep nesting in lock deserialization by delegating field/type handling into focused helper methods, while adding comprehensive tests to ensure persisted lock storage behavior remains stable.

---

## 🚦 CI & Pipeline Health Summary

| Check / Workflow Name | Status | Impact on Review |
| :--- | :--- | :--- |
| `coverage` | ✅ PASSED | All modified lines covered by tests |
| `Pytest (3.14)` | ✅ PASSED | Python 3.14 compatibility confirmed |
| `Autolabel PR` | ⏩ SKIPPED | Benign skip, no impact |
| `Hassfest Validation` | ✅ PASSED | Home Assistant front-end compliance |
| `HACS Validation` | ✅ PASSED | HACS marketplace validation |
| `Autolabel PR` | ✅ PASSED | Autolabeling functional |

> [!NOTE]
> **CI Pipeline Clear:** All GitHub Actions workflows completed successfully.

---.

## 🔍 Architectural Walkthrough

<details>
<summary><b>`custom_components/keymaster/coordinator.py`</b></summary>

* **Mechanical shift:** Extracted the monolithic `_dict_to_kmlocks` method (112 lines) into a clean dispatcher pattern: `._dict_to_kmlocks()` now merely iterates fields and delegates to `._kmlock_field_from_dict()`, returning `cls(**field_values)` at the end.
* **New helpers introduced:**
  - `._kmlock_field_type_from_dict()` – resolves stored type metadata from `keymasterlock_type_lookup`
  - `._kmlock_field_from_dict()` – converts one serialized dataclass field to its runtime value, composing optional unwrapping, temporal parsing, and collection conversion
  - `._kmlock_optional_type_from_dict()` – unwraps `Union` types preserving exactly one non-None type
  - `._kmlock_temporal_value_from_dict()` – converts ISO datetime/time strings with `contextlib.suppress(ValueError)` fallbacks
  - `._kmlock_collection_value_from_dict()` – dispatches to mapping/list handling based on origin type
  - `._kmlock_mapping_value_from_dict()` – converts mapping values **without falling through** to other branches; explicit early return when `len(type_args) != 2`, and separate dataclass-vs-plain-paths with digit-key coercion via `._kmlock_key_from_dict()`
  - `._kmlock_list_value_from_dict()` – recurses only for dataclass item types, preserving plain items
* **Mapping fall-through guard:** The new `_kmlock_mapping_value_from_dict()` explicitly prevents the old code's implicit fall-through by early-returning `field_value` when `len(type_args) != 2`, and then branching dataclass vs. plain-value conversion paths.
* **Tests:** 13+ new test cases in `tests/test_coordinator.py` cover temporal parsing, mapping key coercion, Union unwrapping rules, list conversion, non-dict value preservation, round-trip serialization/deserialization, and edge cases with malformed datetime strings.

</details>

---.

## 🔗 Related / Outside-Diff Context

* **Copilot review (PR #776 #pullrequestreview-5134711297):** Previously raised concern about mapping-to-dataclass conversion potentially raising on malformed persisted values. The author addressed this by:
  - Adding `test_dict_to_kmlocks_preserves_malformed_datetime_string` – malformed temporal strings remain unchanged instead of raising
  - Adding `test_dict_to_kmlocks_mapping_non_dict_value_does_not_fall_through` – mapping-typed non-dict values are not processed as dataclasses or lists
  - Adding `test_dict_to_kmlocks_mapping_without_type_args_preserves_value` – mapping fields without key/value type args pass through unchanged
  - All three tests pass with the refactored code
* **No other modified files:** Only `custom_components/keymaster/coordinator.py` and `tests/test_coordinator.py` were changed, keeping the diff focused and reviewable.

---

All CI checks pass, the author has addressed prior feedback, and the refactoring improves code maintainability without introducing risk. Ready for merge.

**Review by firstof9 — APPROVED** [review]
Looks good.
